### PR TITLE
Implementation of transform chain

### PR DESCRIPTION
**Pull request by pdavis68**
Implements the transform chain to provide more powerful logic options.

I suspect you'll want changes and I've surely missed something, but I wanted it in checked in while it was working (I had it working last night and then broke it and that cost me some hours). It seems to run properly and it saves to the DB properly.

### CODE REVIEW

**Comment by pdavis68** [comment]
Updated to fix the styling of Output Keys and sizing of the textarea for the code.

**Review by acruikshank-wm — CHANGES_REQUESTED** [review]
I think there was some miscommunication. LangChain already has a TransformerChain. We don't need to create our own version (but we do need the TransformSpec and the designer.

It would be ideal if we could rename all the 'Transform' names to 'Transformer' since that's what the LangChain Chain is called.

The rest looks good though.

File: app/src/components/Designer.tsx [hunk 1]
```diff
@@ -283,6 +284,69 @@ const ReformatSpecDesigner = ({ spec }: ReformatSpecDesignerProps) => {
...
+  const [outputKeys, setOutputKeys] = useState<string[]>(spec.output_keys);
+  const [variables, setVariables] = useState<string[]>([]);
+
+
+  useEffect(() => {
+    setTransformFunc(spec.transform_func);
+  }, [spec]);
+
+  const updateTransformFunc = useCallback((value: string) => {
+    setTransformFunc(value);
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply app/src/components/Designer.tsx]: Fixed

File: app/src/components/Designer.tsx [hunk 2]
```diff
@@ -283,6 +284,69 @@ const ReformatSpecDesigner = ({ spec }: ReformatSpecDesignerProps) => {
...
+  }, [transformFunc]);
+
+  const formatReducer = useMemo(() => new FormatReducer(
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply app/src/components/Designer.tsx]: Fixed

File: app/src/components/Designer.tsx [hunk 3]
```diff
@@ -283,6 +284,69 @@ const ReformatSpecDesigner = ({ spec }: ReformatSpecDesignerProps) => {
...
+    <div className="transform-spec spec-designer">
+      <h3 className="chain-id">Transform {spec.chain_id}</h3>
+      <DeleteChainButton onDelete={() => deleteChainSpec(spec.chain_id)}/>
+      <div className="transformfuncs">
+        <div className="form-element">
+          <label>Input Keys</label>
+          <input className="var-name-input" value={inputKeys} onChange={e => setInputKeys(e.target.value.split(","))} placeholder="input keys" />
+        </div>
+        <div className="spec-designer">
+          <HighlightedTextarea
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply app/src/components/Designer.tsx]: Changed to a textarea

File: app/src/components/style/Designer.css [hunk 4]
```diff
@@ -100,7 +100,7 @@
   color: var(--standard-error-color);
 }
 
-.form-element .var-name-input, .form-element .text-input, .form-element textarea {
+.form-element .var-name-input, .var-name-output, .form-element .text-input, .form-element textarea {
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply app/src/components/style/Designer.css]: Remvoed

File: lib/chains/transform_chain.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import logging
+from typing import Dict, List
+from langchain.chains.base import Chain
+from lib.formatters.extended_formatter import ExtendedFormatter
+
+class TransformChain(Chain):
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply lib/chains/transform_chain.py]: I misunderstood how this all worked. I used the ReformatChain as a model, but probably should have used the LLM as a model. I was wondering how langchain was executing the code. I figured I was missing some sort of python behind the scenes voodoo.

File: lib/model/chain_spec.py [hunk 6]
```diff
@@ -122,6 +123,16 @@ def to_lang_chain(self, ctx: LangChainContext) -> Chain:
...
 
 
+class TransformSpec(BaseSpec):
+  chain_type: Literal["transform_spec"]
+  transform_func: str
+  input_keys: List[str]
+  output_keys: List[str]
+
+  def to_lang_chain(self, ctx: LangChainContext) -> Chain:
+    return TransformChain(transform_func=self.transform_func, input_variables=self.input_keys, output_variables=self.output_keys)
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply lib/model/chain_spec.py]: Used the Langchain TransformChain.

**Review by acruikshank-wm — APPROVED** [review]
Nice work.

I have one suggestion about the initializer for the transform spec.

File: app/src/model/spec_control.ts [hunk 7]
```diff
@@ -150,6 +153,13 @@ export const generateDefaultSpec = (type: string): Partial<ChainSpec> => {
         input_keys: [],
         formatters: {'output_key_0': ''},
       };
+    case 'transform_spec':
+      return {
+        chain_type: "transform_spec",
+        transform_func: "",
```
[the remark this answers is not in the document]
Comment (pdavis68) [reply app/src/model/spec_control.ts]: I like that. Done.

**Pull request merged**